File: informatics_custom_apps/eth/page/power_plant_dashboard/power_plant_dashboard.js
frappe.pages["power-plant-dashboard"].on_page_load = function (wrapper) {

    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: "Power Plant Dashboard",
        single_column: true
    });

    page.set_primary_action("Refresh", () => {
        loadDashboard();
    });

    $(page.body).html(`
        <div class="row mb-3">

            <div class="col-md-3">
                <label><b>Date</b></label>
                <input type="date" class="form-control" id="dashboard-date">
            </div>

            <div class="col-md-2">
                <br>
                <button class="btn btn-primary" id="refresh-dashboard">
                    Refresh
                </button>
            </div>

        </div>

        <hr>

        <h4>Daily Plant Readings</h4>

        <div id="plant-summary"></div>

        <hr>

        <h4>Trend Analysis</h4>

        <div id="trend-section">

            <div class="row">

                <div class="col-md-2">
                    <label>From Date</label>
                    <input type="date" id="from-date" class="form-control">
                </div>

                <div class="col-md-2">
                    <label>To Date</label>
                    <input type="date" id="to-date" class="form-control">
                </div>

                <div class="col-md-3">
                    <label>Plant</label>
                    <select id="trend-plant" class="form-control"></select>
                </div>

                <div class="col-md-3">
                    <label>Parameter</label>
                    <select id="trend-parameter" class="form-control"></select>
                </div>

                <div class="col-md-2">
                <div class="form-check" style="margin-top:32px;">
                    <input
                        class="form-check-input"
                        type="checkbox"
                        id="show-limits"
                    >
                    <label class="form-check-label" for="show-limits">
                        Show Limits
                    </label>
                </div>
                </div>

                <div class="col-md-2">
                    <br>
                    <button class="btn btn-primary" id="show-trend">
                        Show Trend
                    </button>
                </div>

            </div>

            <br>

            <div id="trend-chart"></div>

        </div>
    `);

    $("#dashboard-date").val(frappe.datetime.get_today());

    loadDashboard();

    $("#refresh-dashboard").click(loadDashboard);

$(page.body).on("click", "#show-trend", function () {
    console.log("Show Trend Clicked");

    let from_date = $("#from-date").val();
    let to_date = $("#to-date").val();
    let plant = $("#trend-plant").val();
    let parameter = $("#trend-parameter").val();

    if (!from_date || !to_date || !plant || !parameter) {
        frappe.msgprint("Please select all filters.");
        return;
    }

    frappe.call({
        method: "informatics_custom_apps.eth.page.power_plant_dashboard.power_plant_dashboard.get_parameter_trend",
        args: {
            from_date,
            to_date,
            plant,
            parameter
        },
        callback(r) {
            console.log("Trend Response");
            console.log(r.message);

            renderTrendChart(r.message);
        }
    });
});

}
function populatePlants() {

    const plants = [
        "Buttar Biofuels",
        "RSL Belwara",
        "RSL Buttar",
        "RSL Louhka",
        "RSLD Karnal",
        "Superior Biofuels"
    ];

    let options = '<option value="">Select Plant</option>';

    plants.forEach(function(plant) {
        options += `<option value="${plant}">${plant}</option>`;
    });

    $("#trend-plant").html(options);
}
function loadParameters() {

    frappe.call({
        method:
        "informatics_custom_apps.eth.page.power_plant_dashboard.power_plant_dashboard.get_parameters",

        callback(r) {

            let options = '<option value="">Select Parameter</option>';

            r.message.forEach(p => {

                options += `
                <option value="${p.fieldname}">
                    ${p.section ? p.section + " → " : ""}${p.label}
                </option>`;
            });

            $("#trend-parameter").html(options);
        }
    });

}
function loadDashboard() {

    frappe.call({
        method: "informatics_custom_apps.eth.page.power_plant_dashboard.power_plant_dashboard.get_dashboard",

        args: {
            date: $("#dashboard-date").val()
        },

        callback(r) {

            console.log("========== DASHBOARD ==========");
            console.table(r.message.plants);

            renderPlantSummary(r.message.plants);

            // Populate trend filters
            populatePlants();
            loadParameters();
        }
    });

}
function renderPlantSummary(plants) {
    console.log("Plants received by renderer:");
    console.table(plants);
    let html = `
        <table class="table table-bordered table-hover">

            <thead>

                <tr>
                    <th width="40"></th>
                    <th>Plant</th>
                    <th>Violations</th>
                    <th>Last Updated</th>
                </tr>

            </thead>

            <tbody>
    `;

    plants.forEach(p => {

        let badge = "success";

        if (p.violations > 0)
            badge = "danger";

        html += `

            <tr class="plant-row"
                data-plant="${p.plant}">

                <td>▶</td>

                <td><b>${p.plant}</b></td>

                <td>

                    <span class="badge badge-${badge}">
                        ${p.violations}
                    </span>

                </td>

                <td>${p.last_updated || ""}</td>

            </tr>

            <tr
                id="details-${frappe.scrub(p.plant)}"
                style="display:none;">

                <td colspan="4">

                    <div class="plant-details">
                        Loading...
                    </div>

                </td>

            </tr>
        `;

    });

    html += "</tbody></table>";

    $("#plant-summary").html(html);

}
$(document).on("click", ".plant-row", function () {

    let plant = $(this).data("plant");

    let id = "#details-" + frappe.scrub(plant);

    if ($(id).is(":visible")) {

        $(id).hide();

        return;
    }

    $(".plant-details").parent().parent().hide();

    $(id).show();

    loadPlantDetails(
        plant,
        id + " .plant-details"
    );

});
function loadPlantDetails(plant, target) {

    $(target).html("Loading...");

    frappe.call({

        method:
        "informatics_custom_apps.eth.page.power_plant_dashboard.power_plant_dashboard.get_plant_logbook",

        args: {

            plant: plant,

            date: $("#dashboard-date").val()

        },

        callback(r) {

            console.log("Plant Logbook Response");
    console.log(r.message);

    renderPlantLogbook(target, r.message);

        }

    });

}
function renderPlantLogbook(target, data) {

    if (!data || !data.length) {
        $(target).html(`
            <div class="alert alert-warning">
                No logbook data found.
            </div>
        `);
        return;
    }

    const slots = [
        "9:30 AM",
        "1:30 PM",
        "5:30 PM",
        "9:30 PM",
        "1:30 AM",
        "5:30 AM"
    ];

    let html = `
        <table class="table table-bordered table-sm">

            <thead>

                <tr>

                    <th>Parameter</th>

                    <th>Unit</th>
    `;

    slots.forEach(slot => {
        html += `<th>${slot}</th>`;
    });

    html += `
                    <th style="background:#eef5ff;">Min</th>
                    <th style="background:#eef5ff;">Max</th>

                </tr>

            </thead>

            <tbody>
    `;

    let current_section = "";

    data.forEach(p => {

        if (p.section !== current_section) {

            current_section = p.section;

            html += `
                <tr class="table-secondary">
                    <th colspan="${slots.length + 4}">
                        ${current_section || "Other"}
                    </th>
                </tr>
            `;
        }

        html += `<tr>`;

        html += `<td>${p.label}</td>`;

        html += `<td>${p.unit || ""}</td>`;

        slots.forEach(slot => {

            let cell = p.values[slot] || {};

            let value = cell.value ?? "";

            let status = cell.status || "";

            let style = "";

            if (status === "High") {
                style = 'background:#f8d7da;';
            }
            else if (status === "Low") {
                style = 'background:#fff3cd;';
            }

            html += `
                <td style="${style}">
                    ${value}
                </td>
            `;
        });

        html += `
            <td style="background:#eef5ff;font-weight:600;">
                ${p.min ?? "-"}
            </td>

            <td style="background:#eef5ff;font-weight:600;">
                ${p.max ?? "-"}
            </td>
        `;

        html += `</tr>`;
    });

    html += `
            </tbody>
        </table>
    `;

    $(target).html(html);

}


function renderTrendChart(result) {
    if (!result.data.length) {

        $("#trend-chart").html(`
            <div class="alert alert-warning">
                No data found.
            </div>
        `);

        return;
    }

    const hasMin = result.norm &&
        result.norm.min !== null &&
        result.norm.min !== "" &&
        result.norm.min !== 0;

    const hasMax = result.norm &&
        result.norm.max !== null &&
        result.norm.max !== "" &&
        result.norm.max !== 0;

    let labels = [];
    let values = [];
    let minValues = [];
    let maxValues = [];

    let normal = 0;
    let low = 0;
    let high = 0;

    result.data.forEach(d => {

        labels.push(`${d.date} ${d.time_slot}`);
        values.push(d.value);

        if (hasMin) {
            minValues.push(result.norm.min);
        }

        if (hasMax) {
            maxValues.push(result.norm.max);
        }

        if (hasMin && d.value < result.norm.min) {
            low++;
        }
        else if (hasMax && d.value > result.norm.max) {
            high++;
        }
        else {
            normal++;
        }

    });

    let rangeText = "";

    if (hasMin || hasMax) {

        rangeText = `
            <b>Normal Range :</b>
            ${hasMin ? result.norm.min : "-"}
            -
            ${hasMax ? result.norm.max : "-"}
            ${result.norm.unit || ""}
        `;

    } else {

        rangeText = `
            <b>Normal Range :</b> Not Configured
        `;

    }

    $("#trend-chart").html(`
        <div style="margin-bottom:15px;">

            <div><b>Plant :</b> ${$("#trend-plant").val()}</div>

            <div>
                <b>Parameter :</b>
                ${$("#trend-parameter option:selected").text()}
            </div>

            <div>${rangeText}</div>

            <br>

            <span class="badge badge-success">
                Normal : ${normal}
            </span>

            <span class="badge badge-warning">
                Low : ${low}
            </span>

            <span class="badge badge-danger">
                High : ${high}
            </span>

        </div>
    <div id="trend-graph"></div>
    `);

    const showLimits = $("#show-limits").is(":checked");

    let datasets = [
        {
            name: "Actual",
            values: values
        }
    ];

    if (showLimits && hasMin) {
        datasets.push({
            name: "Min",
            values: minValues
        });
    }

    if (showLimits && hasMax) {
        datasets.push({
            name: "Max",
            values: maxValues
        });
    }

    new frappe.Chart("#trend-graph", {

        title: `${$("#trend-parameter option:selected").text()} Trend`,

        data: {
            labels: labels,
            datasets: datasets
        },

        type: "line",

        height: 350,

        lineOptions: {
            hideDots: false,
            regionFill: false
        },
        axisOptions: {
            min: 0
        }

    });
    let table = `
    <br>
   <details style="margin-top:20px;">
    <summary><b>Recorded Values (${result.data.length})</b></summary>

    <table class="table table-bordered table-sm table-hover">
        <thead>
            <tr>
                <th>Date</th>
                <th>Time Slot</th>
                <th>Value</th>
                <th>Status</th>
            </tr>
        </thead>
        <tbody>
        </details>
`;

result.data.forEach(d => {

    let status = "Normal";
    let badge = "success";

    if (hasMin && d.value < result.norm.min) {
        status = "Low";
        badge = "warning";
    }
    else if (hasMax && d.value > result.norm.max) {
        status = "High";
        badge = "danger";
    }

    table += `
        <tr>
            <td>${d.date}</td>
            <td>${d.time_slot}</td>
            <td>
                <b>${d.value}</b>
                ${result.norm.unit || ""}
            </td>
            <td>
                <span class="badge badge-${badge}">
                    ${status}
                </span>
            </td>
        </tr>
    `;
});

table += `
        </tbody>
    </table>
`;

$("#trend-chart").append(table);

}

File: informatics_custom_apps/eth/page/power_plant_dashboard/power_plant_dashboard.py
from unittest import result
import pprint
import frappe

@frappe.whitelist()
def get_dashboard(date=None):

    if not date:
        date = frappe.utils.today()

    logbooks = frappe.get_all(
        "Power Plant Log Book",
        filters={"log_date": date},
        fields=["name", "plant", "log_date"],
        order_by="plant"
    )

    norms = get_norms()

    plants = []

    for log in logbooks:

        violation_count = 0

        doc = frappe.get_doc("Power Plant Log Book", log.name)

        for row in doc.logs:

            for fieldname, norm in norms.items():

                value = row.get(fieldname)

                if value in (None, "", 0):
                    continue

                status = get_status(
                    value,
                    norm["min"],
                    norm["max"]
                )

                if status in ("Low", "High"):
                    violation_count += 1

        plants.append({
            "name": log.name,
            "plant": log.plant,
            "last_updated": log.log_date,
            "violations": violation_count
        })

    return {
        "plants": plants
    }

def get_norms():

    doc = frappe.get_single("ETH Logbook Norms")

    norms = {}

    for row in doc.power_plant_log_norms:

        if not row.fieldname:
            continue

        norms[row.fieldname] = {
            "section": row.section,
            "min": row.min_value,
            "max": row.max_value,
            "unit": row.unit or ""
        }

    return norms

@frappe.whitelist()
def get_parameters():

    meta = frappe.get_meta("Power Plant Log Book Item")

    parameters = []

    for df in meta.fields:

        if df.fieldtype in ("Float", "Currency", "Int"):

            parameters.append({
                "fieldname": df.fieldname,
                "label": df.label
            })

    parameters.sort(key=lambda x: x["label"])

    return parameters

# def calculate_plant_summary(norms, filters=None):

#     logbooks = get_logs(filters or {})

#     log_items = get_log_items([d.name for d in logbooks])

#     plant_summary = {}

#     for log in logbooks:

#         if log.plant not in plant_summary:
#             plant_summary[log.plant] = {
#                 "plant": log.plant,
#                 "normal": 0,
#                 "low": 0,
#                 "high": 0,
#                 "last_updated": log.log_date,
#             }

#         summary = plant_summary[log.plant]

#         # Keep latest log date
#         if log.log_date > summary["last_updated"]:
#             summary["last_updated"] = log.log_date

#         rows = log_items.get(log.name, [])

#         for row in rows:

#             for fieldname, norm in norms.items():

#                 value = row.get(fieldname)

#                 if value in (None, "", 0, 0.0):
#                     continue

#                 status = get_status(
#                     value,
#                     norm["min"],
#                     norm["max"]
#                 )

#                 if status == "Low":
#                     summary["low"] += 1

#                 elif status == "High":
#                     summary["high"] += 1

#                 else:
#                     summary["normal"] += 1

#     plants = []

#     for summary in plant_summary.values():

#         total = (
#             summary["normal"]
#             + summary["low"]
#             + summary["high"]
#         )

#         summary["total"] = total

#         summary["health"] = (
#             round((summary["normal"] / total) * 100, 2)
#             if total else 100
#         )

#         plants.append(summary)

#     plants.sort(key=lambda x: x["health"])

#     return plants

def get_field_labels():

    meta = frappe.get_meta("Power Plant Log Book Item")

    labels = {}

    for df in meta.fields:

        labels[df.fieldname] = df.label

    return labels

# def get_logs(filters):

#     conditions = {}

#     if filters.get("from_date"):
#         conditions["log_date"] = [">=", filters.from_date]

#     if filters.get("to_date"):

#         if "log_date" in conditions:
#             conditions["log_date"] = [
#                 "between",
#                 [filters.from_date, filters.to_date]
#             ]
#         else:
#             conditions["log_date"] = ["<=", filters.to_date]

#     if filters.get("plant"):
#         conditions["plant"] = filters.plant

#     return frappe.get_all(
#         "Power Plant Log Book",
#         filters=conditions,
#         fields=[
#             "name",
#             "plant",
#             "log_date"
#         ],
#         order_by="log_date asc"
#      )

# def get_log_items(parents):

#     if not parents: 
#         return {}

#     rows = frappe.get_all(
#         "Power Plant Log Book Item",
#         filters={
#             "parent": ["in", parents]
#         },
#         fields=["*"],
#         order_by="parent asc, idx asc"
#     )

#     grouped = {}

#     for row in rows:
#         grouped.setdefault(row.parent, []).append(row)

#     return grouped

def get_status(value, min_value=None, max_value=None):

    if value in (None, "", 0, 0.0):
        return None

    # Treat blank norms (stored as 0) as not configured
    has_min = min_value not in (None, "", 0, 0.0)
    has_max = max_value not in (None, "", 0, 0.0)

    if has_min and value < min_value:
        return "Low"

    if has_max and value > max_value:
        return "High"

    # If no norm exists at all, don't assign any status
    if not has_min and not has_max:
        return None

    return "Normal"

@frappe.whitelist()
def get_plant_logbook(plant, date):

    norms = get_norms()
    labels = get_field_labels()

    log = frappe.get_value(
        "Power Plant Log Book",
        {
            "plant": plant,
            "log_date": date
        },
        "name"
    )

    if not log:
        return []

    rows = frappe.get_all(
        "Power Plant Log Book Item",
        filters={"parent": log},
        fields=["*"],
        order_by="idx"
    )

    meta = frappe.get_meta("Power Plant Log Book Item")

    skip_fields = (
        "name",
        "owner",
        "creation",
        "modified",
        "modified_by",
        "parent",
        "parentfield",
        "parenttype",
        "doctype",
        "idx",
        "time_slot"
    )

    parameters = {}

    for row in rows:

        time_slot = row.time_slot

        for df in meta.fields:

            if df.fieldtype not in ("Float", "Int", "Currency"):
                continue

            fieldname = df.fieldname

            if fieldname in skip_fields:
                continue

            if fieldname not in parameters:

                norm = norms.get(fieldname, {})

                parameters[fieldname] = {
                    "fieldname": fieldname,
                    "label": labels.get(fieldname, fieldname),
                    "section": norm.get("section"),
                    "unit": norm.get("unit", ""),
                    "min": norm.get("min"),
                    "max": norm.get("max"),
                    "values": {}
                }

            value = row.get(fieldname)

            if value in (None, 0, 0.0):
                value = ""

            status = ""

            norm = norms.get(fieldname)

            if norm and value != "":
                status = get_status(
                    value,
                    norm.get("min"),
                    norm.get("max")
                ) or ""

            parameters[fieldname]["values"][time_slot] = {
                "value": value,
                "status": status
            }

    return list(parameters.values())


@frappe.whitelist()
def get_parameter_trend(plant, parameter, from_date, to_date):

    norms = get_norms()

    logs = frappe.get_all(
        "Power Plant Log Book",
        filters={
            "plant": plant,
            "log_date": ["between", [from_date, to_date]]
        },
        fields=["name", "log_date"],
        order_by="log_date"
    )

    result = []

    for log in logs:

        doc = frappe.get_doc("Power Plant Log Book", log.name)

        for row in doc.logs:

            value = row.get(parameter)

            if value in (None, "", 0):
                continue

            result.append({
                "date": str(log.log_date),
                "time_slot": row.time_slot,
                "value": value
            })

    norm = norms.get(parameter)

    return {
        "data": result,
        "norm": norm
    }